File: removeElement.js
/**
 * @param {number[]} nums
 * @param {number} val
 * @return {number}
 */
var removeElement = function (nums, val) {
    let index;
    let count = 0;
    for (index = 0; index < nums.length; index++) {
        if (nums[index] == val) {
            shiftElm(index, nums);
            count++;
            index--;
        }
    }
    return index - count;
};

function shiftElm(index, nums) {
    for (let i = index; i < nums.length; i++) {
        nums[i] = nums[i + 1];
    }
}

let nums = [3, 2, 2, 3];
let res = removeElement(nums, 3);
console.log(res);

// find the matching val in nums
// then shift the rest of the nums to the left by the amount of 1

File: removeElement.c
#include<stdio.h>
#define N 6

void shift(int* nums, int index, int numsSize) {
  for (int i = index; i < numsSize; i++) {
    int replace;
    if(i + 1 == numsSize) {
      replace = -1;
    } else {
      replace = nums[i + 1];
    }
    nums[i] = replace;
  }
}

int removeElement(int* nums, int numsSize, int val){
  int elmWithoutVal = numsSize;
  for (int i = 0; i < numsSize; i++) {
    if(nums[i] == val) {
      elmWithoutVal = elmWithoutVal - 1;
      shift(nums, i, numsSize);
      i = i - 1;
    }
  }

  return elmWithoutVal;
}


int main()
{
  int a[N] = {4,4,0,1,0,2};
  int ret = removeElement(a, N, 0);
  printf("%d\n\n", ret);

  for (int i = 0; i < N; i++) {
    printf("%d\n", a[i]);
  }


  return 0;
}
